Framework/API/test/DetectorSearcherTest.h

```diff
@@ -0,0 +1,235 @@
+
+    checkResult(V3D(-0.959758, -0, 0.280828), 5);
+    checkResult(V3D(0.913156, -0.285361, 0.291059), 6);
+    checkResult(V3D(-6.09343e-17, -0.995133, 0.0985376), 7);
+    checkResult(V3D(-0.913156, -0.285361, 0.291059), 8);
+    checkResult(V3D(0.942022, 0.294382, 0.161038), 9);
+    checkResult(V3D(-6.11563e-17, 0.998759, 0.0498137), 10);
+    checkResult(V3D(-0.942022, 0.294382, 0.161038), 11);
+    checkResult(V3D(0.988034, -1.20999e-16, 0.154233), 12);
+
+    checkResult(V3D(-0.988034, -0, 0.154233), 14);
+    checkResult(V3D(0.942022, -0.294382, 0.161038), 15);
+    checkResult(V3D(-6.11563e-17, -0.998759, 0.0498137), 16);
+    checkResult(V3D(-0.942022, -0.294382, 0.161038), 17);
+    checkResult(V3D(0.948717, 0.296474, 0.109725), 18);
+    checkResult(V3D(-6.11984e-17, 0.999446, 0.0332779), 19);
+    checkResult(V3D(-0.948717, 0.296474, 0.109725), 20);
+    checkResult(V3D(0.994483, -1.21789e-16, 0.104898), 21);
+
+    checkResult(V3D(-0.994483, -0, 0.104898), 23);
+    checkResult(V3D(0.948717, -0.296474, 0.109725), 24);
+    checkResult(V3D(-6.11984e-17, -0.999446, 0.0332779), 25);
+    checkResult(V3D(-0.948717, -0.296474, 0.109725), 26);
+  }
+
+  void test_invalid_rectangular() {
+    auto inst =
+        ComponentCreationHelper::createTestInstrumentRectangular2(1, 100);
+
+    ExperimentInfo expInfo;
+    expInfo.setInstrument(inst);
+    const auto &info = expInfo.detectorInfo();
+
+    DetectorSearcher searcher(inst, info);
+    const auto resultNull = searcher.findDetectorIndex(V3D(0, 0, 0));
+    TS_ASSERT(!std::get<0>(resultNull))
+
+    const auto resultNaN = searcher.findDetectorIndex(V3D(NAN, NAN, NAN));
+    TS_ASSERT(!std::get<0>(resultNaN))
+  }
+
+  void test_invalid_cylindrical() {
+    auto inst = ComponentCreationHelper::createTestInstrumentCylindrical(
+        3, V3D(0, 0, -1), V3D(0, 0, 0), 1.6, 1.0);
+    ExperimentInfo expInfo;
+    expInfo.setInstrument(inst);
+    const auto &info = expInfo.detectorInfo();
+
+    DetectorSearcher searcher(inst, info);
+    const auto resultNull = searcher.findDetectorIndex(V3D(0, 0, 0));
+    TS_ASSERT(!std::get<0>(resultNull))
+
+    const auto resultNaN = searcher.findDetectorIndex(V3D(NAN, NAN, NAN));
+    TS_ASSERT(!std::get<0>(resultNaN))
+  }
+
+  void test_search_rectangular() {
+    auto inst =
+        ComponentCreationHelper::createTestInstrumentRectangular2(1, 100);
+    ExperimentInfo expInfo;
+    expInfo.setInstrument(inst);
+    const auto &info = expInfo.detectorInfo();
+
+    DetectorSearcher searcher(inst, info);
+    const auto checkResult = [&searcher](V3D q, size_t index) {
+      const auto result = searcher.findDetectorIndex(q);
+      TS_ASSERT(std::get<0>(result))
+      TS_ASSERT_EQUALS(std::get<1>(result), index)
+    };
+
+    for (size_t pointNo = 0; pointNo < info.size(); ++pointNo) {
+      const auto &det = info.detector(pointNo);
+      const auto q = convertDetectorPositionToQ(det);
+      checkResult(q, pointNo);
```

Comment:
No. For each detector it converts the detector position to q, then searches for the detector using that q. We should get the same detector index back because that q should perfectly hit the detector.